### PR TITLE
[SPARK-34334][K8S] Correctly identify timed out pending pod requests as excess request

### PR DESCRIPTION
**Pull request by attilapiros**
### What changes were proposed in this pull request?

Fixing identification of timed-out pending pod requests as excess requests to delete when the excess is higher than the newly created timed out requests and there is some non-timed out newly created requests too.

### Why are the changes needed?

After https://github.com/apache/spark/pull/29981 only timed out newly created requests and timed out pending requests are taken as excess request. 

But there is small bug when the excess is higher than the newly created timed out requests and there is some non-timed out newly created requests as well. Because all the newly created requests are counted as excess request when items are chosen from the timed out pod pending requests.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?
 
 There is new unit test added: `SPARK-34334: correctly identify timed out pending pod requests as excess`.

### CODE REVIEW

**Comment by attilapiros** [comment]
cc @dongjoon-hyun @HyukjinKwon @holdenk

**Comment by holdenk** [comment]
Thanks for working on this :)

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala [hunk 1]
```diff
@@ -223,14 +223,15 @@ private[spark] class ExecutorPodsAllocator(
...
         val excess = knownPodCount - targetNum
+        val newlyCreatedToDelete = newlyCreatedExecutorsForRpId
+          .filter { case (_, (_, createTime)) =>
+            currentTime - createTime > executorIdleTimeout
+          }.keys.take(excess).toList
         val knownPendingToDelete = currentPendingExecutors
           .filter(x => isExecutorIdleTimedOut(x._2, currentTime))
+          .take(excess - newlyCreatedToDelete.size)
           .map { case (id, _) => id }
-          .take(excess - newlyCreatedExecutorsForRpId.size)
```
Comment (attilapiros) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsAllocator.scala:229]: Before this PR this line caused the error in the new unit test.

As some `newlyCreatedExecutorsForRpId` might be not timed out and in place of those we should choose timed out pending executor pod requests.

**Comment by attilapiros** [comment]
jenkins retest this please

**Comment by holdenk** [comment]
LGTM pending CI.
Jenkins failure is unrelated but to be safe, Jenkins retest this please.

**Comment by attilapiros** [comment]
jenkins retest this please

**Comment by holdenk** [comment]
Out of disk space error on the Jenkins machine is unrelated (`failed to copy files: failed to copy directory: Error processing tar file(exit status 1): write /hadoop-client-api-3.2.2.jar: no space left on device`), I'm going to merge this bug fix.

**Comment by holdenk** [comment]
Merged to the current dev & 3.1 branches.

**Comment by attilapiros** [comment]
Thanks @holdenk!

**Pull request closed without merging**